network/src/metrics.rs

```diff
@@ -1,155 +1,150 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use prometheus::{default_registry, register_int_gauge_vec_with_registry, IntGaugeVec, Registry};
+use anemo_tower::callback::MakeCallbackHandler;
+use anemo_tower::callback::ResponseHandler;
+use prometheus::{
+    register_histogram_vec_with_registry, register_int_counter_vec_with_registry,
+    register_int_gauge_vec_with_registry, IntGaugeVec, Registry,
+};
+use prometheus::{HistogramVec, IntCounterVec};
 use std::sync::Arc;
-
-pub trait NetworkMetrics {
-    fn network_available_tasks(&self) -> &IntGaugeVec;
+use std::time::Instant;
+
+#[derive(Clone)]
+pub struct NetworkMetrics1 {
+    /// Counter of requests, route is a label (ie separate timeseries per route)
+    requests: IntCounterVec,
+    /// Request latency, route is a label
+    request_latency: HistogramVec,
+    concurrent_requests: IntGaugeVec,
+    /// Failed requests by route
+    errors: IntCounterVec,
 }
 
-#[derive(Clone, Debug)]
-pub struct PrimaryNetworkMetrics {
-    /// The number of executor available tasks
-    pub network_available_tasks: IntGaugeVec,
-}
+const LATENCY_SEC_BUCKETS: &[f64] = &[
+    0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5, 1., 2.5, 5., 10., 20., 30., 60., 90.,
+];
+
+impl NetworkMetrics1 {
+    pub fn new(node: &'static str, registry: &Registry) -> Self {
+        let requests = register_int_counter_vec_with_registry!(
+            format!("{node}_requests"),
+            "The number of anemo requests made on the network",
+            &["direction", "route"],
+            registry
+        )
+        .unwrap();
+
+        let request_latency = register_histogram_vec_with_registry!(
+            format!("{node}_requests_latency"),
+            "Latency of a request by route",
+            &["direction", "route"],
+            LATENCY_SEC_BUCKETS.to_vec(),
+            registry,
+        )
+        .unwrap();
+
+        let concurrent_requests = register_int_gauge_vec_with_registry!(
+            format!("{node}_concurrent_requests"),
+            "The number of concurrent network requests",
+            &["direction", "route"],
+            registry
+        )
+        .unwrap();
+
+        let errors = register_int_counter_vec_with_registry!(
+            format!("{node}_errors"),
```

Comment:
`{node}_request_errors`? Just cause I can imagine that name being too generic and confusing with respect to other error metrics we might have?

Reply:
good catch